src/System.Text.Primitives/System/Text/InvariantParser_byte.cs

```diff
@@ -77,40 +119,87 @@ unsafe public static bool TryParse(byte* utf8Text, int index, int length, out by
             bytesConsumed = 0;
 
-            for (int byteIndex = index; byteIndex < length + index; byteIndex++)
+            if (cultureAndEncodingInfo.IsInvariantUtf8)
             {
-                byte nextByteVal = (byte)(utf8Text[byteIndex] - '0');
-                if (nextByteVal > 9) // if the next character is not a digit
+                for (int byteIndex = index; byteIndex < length + index; byteIndex++)
                 {
-                    if (bytesConsumed == 0)
+                    byte nextByteVal = (byte)(utf8Text[byteIndex] - '0');
+                    if (nextByteVal > 9) // if the next character is not a digit
                     {
-                        value = default(byte);
+                        if (bytesConsumed == 0)
+                        {
+                            value = default(byte);
+                            return false;
+                        }
+                        else
+                        {
+                            return true;
+                        }
+                    }
+                    else if (value > Byte.MaxValue / 10)
+                    {
+                        value = 0;
+                        bytesConsumed = 0;
                         return false;
                     }
-                    else
+                    else if (value > 0 && Byte.MaxValue - value * 10 < nextByteVal) // overflow
                     {
-                        return true;
+                        value = 0;
+                        bytesConsumed = 0;
+                        return false;
                     }
+
+                    byte candidate = (byte)(value * 10 + nextByteVal); // left shift the value and add the nextByte.
+
+                    value = candidate;
+                    bytesConsumed++;
                 }
-                else if (value > Byte.MaxValue / 10)
-                {
-                    value = 0;
-                    bytesConsumed = 0;
-                    return false;
-                }
-                else if (value > 0 && Byte.MaxValue - value * 10 < nextByteVal) // overflow
+            }
+            else if (cultureAndEncodingInfo.IsInvariantUtf16)
+            {
+                for (int byteIndex = index; byteIndex < length + index - 1; byteIndex += 2) // loop through the byte array two bytes at a time for UTF-16
                 {
-                    value = 0;
-                    bytesConsumed = 0;
-                    return false;
-                }
-
-                byte candidate = (byte)(value * 10 + nextByteVal); // left shift the value and add the nextByte.
+                    byte nextByte = utf8Text[byteIndex];
+                    byte byteAfterNext = utf8Text[byteIndex + 1];
+                    byte nextByteVal = (byte)(nextByte - '0');
+                    if (nextByteVal > 9 || byteAfterNext != 0) // if the second byte isn't zero, this isn't an ASCII-equivalent code unit and we can quit here
```

Comment:
can we have just one PR have the whole change? it will be easier than doing it twice.

---
In reply to: [75390658](https://github.com/dotnet/corefxlab/pull/762#discussion_r75390658) [](ancestors = 75390658)